src/dropck.md

```diff
@@ -22,268 +18,152 @@ let y;
+This is the *sound generic drop* issue. We call it that because it only applies
+to destructors of generic types. That is, if `Inspector` weren't generic, it
+couldn't store any lifetime other than `'static', and dangling pointers would
+never be a concern. The enforcement of sound generic drop is handled by the
+*drop check*, which is more commonly known as *dropck*.
 
-(Future changes to the language, such as impl specialization, may add
-other avenues for such indirect access.)
+It turns out that getting dropck's design exactly right has been very difficult.
+This is because, as we'll see, we don't want it to give up completely on generic
+destructors. In particular: what if we knew `Inspector` *didn't* or even *couldn't*
+dereference the pointer in its destructor? Wouldn't it be nice if that meant our
+code compiled again?
 
-Here is an example of invoking a callback:
+Since Rust 1.0, and as of Rust 1.18, there have been two changes to how dropck
+works due to soundness issues. At least one more is planned, as the latest
+version was intended to be a temporary hack.
 
-```rust,ignore
-struct Inspector<T>(T, &'static str, Box<for <'r> fn(&'r T) -> String>);
+* [non-parametric dropck](https://github.com/rust-lang/rfcs/blob/master/text/1238-nonparametric-dropck.md)
+* [dropck eyepatch](https://github.com/rust-lang/rfcs/blob/master/text/1327-dropck-param-eyepatch.md)
 
-impl<T> Drop for Inspector<T> {
-    fn drop(&mut self) {
-        // The `self.2` call could access a borrow e.g. if `T` is `&'a _`.
-        println!("Inspector({}, {}) unwittingly inspects expired data.",
-                 (self.2)(&self.0), self.1);
-    }
-}
-```
-
-Here is an example of a trait method call:
+Old versions of this document were based on the original 1.0 design, which
+was unsafe by default, and therefore expected unsafe Rust programmers to manually
+opt out of it.
 
-```rust,ignore
-use std::fmt;
+The good news is that, unlike the 1.0 design, the 1.18 design is *safe by default*:
+you can completely ignore that dropck exists, and nothing bad can be done with
+your types. But sometimes you will be able to leverage your knowledge of dropck
+to make transiently dangling references work, and that's nice.
 
-struct Inspector<T: fmt::Display>(T, &'static str);
+So here's all you should need to know about dropck these days:
 
-impl<T: fmt::Display> Drop for Inspector<T> {
-    fn drop(&mut self) {
-        // There is a hidden call to `<T as Display>::fmt` below, which
-        // could access a borrow e.g. if `T` is `&'a _`
-        println!("Inspector({}, {}) unwittingly inspects expired data.",
-                 self.0, self.1);
-    }
-}
-```
+* If a type isn't generic, then it cannot contain borrows that expire, and
+  is therefore uninteresting.
+* If a type is generic, and doesn't have a destructor, its generic arguments
+  must must live *at least* as long as it.
+* If a type is generic, and *does* have a destructor, its generic arguments
```

Comment:
Please please no "strictly"/"at least" language. That is (A) has no chance of working with NLL and (B) is a bad way of thinking about things.